---
title: "Exploring Azoreductases in healthy, IBD and CRC individiuals"
author: "Domenick J. Braccia"
date: "`Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Loading Libraries
```{r load libraries, echo=FALSE, message=FALSE, warning=FALSE}
library(curatedMetagenomicData)
library(tictoc)
library(dplyr)
library(ggplot2)
library(VennDiagram)
library(UpSetR)
library(patchwork)
library(cowplot)
library(svglite)
library(scales)
library(reshape2)
library(seqinr)
library(stringr)
library(ggforce)
library(ggpubr)
library(ggridges)
library(lme4)
library(lmerTest)
library(stringr)
```

## Loading and processing data
```{r process data, message=FALSE}
source("process_data_hmm.R")
```

## Load GutFunFind results

```{r get seq hits}
source("gene_hits_table_hmm.R")
```

## Examining RA of azoreducing species in healthy controls

```{r healthy RA}
source("healthy_relative_abundances_hmm.R")
```

## RA of azoreducing species in at risk groups (IBD, CRC)

```{r}
source("risk_relative_abundances.R")
```

## fluctuation of azoreducing genes and species over time (HMP2 only)

```{r figure 5}
source("azored_fluctuation.R")
```

# BASEMENT
remaining code is depreciated. 

## Abundance of azoreducing genes

```{r}
source("gene_abundances.R")
```

## Heterogeneity of azoreducing genes and species across individuals

```{r azored heterogeneity, message=FALSE, warning=FALSE}
setwd("/Users/dome/Desktop/UMD/hall-pop-lab/Azoreductases/")
source("azored_heterogeneity.R")

row1 <- (p1a + theme(legend.position = "none") | p1b + theme(legend.position = "none"))
row2 <- (p2a + theme(legend.position = "none") | p2b + theme(legend.position = "none"))
p3 <- plot_grid(p1a + theme(legend.position = "none"),
                p1b + theme(legend.position = "none"),
                p2a + theme(legend.position = "none"),
                p2b + theme(legend.position = "none"),
                legend_a, legend_b, rel_heights = c(1, 1, 0.2, 0.2), ncol = 2)

# ggarrange(row1,
#           row2,
#           ggarrange())

pA <- plot_grid((p1a + theme(legend.position = "none")) / (p2a + theme(legend.position = "none")), 
               legend_a, ncol = 1, rel_heights = c(1, 0.2))
pB <- plot_grid((p1b + theme(legend.position = "none")) / (p2b + theme(legend.position = "none")), 
               legend_b, ncol = 1, rel_heights = c(1, 0.2))

p <- pA | pB

# ggsave("figures/figure3/azored_heterogeneity.png", plot = p3, width = 7, height = 7)
# ggsave("figures/figure3/azored_heterogeneity.svg", plot = p3, width = 7, height = 7)
```

